Bidirectional communication on config channel

Qbicz opened the pull request:
Specify a frame for communication in both directions over config channel on USB and BLE.

Qbicz commented:
Please hold on with review; there will be one more commit specifying the data frame.

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:705: WARNING:LINE_SPACING: Missing a blank line after declarations
#705: FILE: samples/nrf_desktop/src/services/hids.c:197:
+	int err = frame_size_check(frame, length);
+	if (err) {

-:713: WARNING:LINE_SPACING: Missing a blank line after declarations
#713: FILE: samples/nrf_desktop/src/services/hids.c:202:
+	size_t pos;
+	if (usb) {

-:714: ERROR:C99_COMMENTS: do not use C99 // comments
#714: FILE: samples/nrf_desktop/src/services/hids.c:203:
+		// TODO: parse REPORT ID on USB

-:740: WARNING:LONG_LINE_COMMENT: line over 80 characters
#740: FILE: samples/nrf_desktop/src/services/hids.c:221:
+	// TODO: check when you have USB implemented if its better to copy pointer or not

-:740: ERROR:C99_COMMENTS: do not use C99 // comments
#740: FILE: samples/nrf_desktop/src/services/hids.c:221:
+	// TODO: check when you have USB implemented if its better to copy pointer or not

-:749: WARNING:LINE_SPACING: Missing a blank line after declarations
#749: FILE: samples/nrf_desktop/src/services/hids.c:230:
+	int err = frame_size_check(frame, length);
+	if (err) {

-:755: WARNING:LINE_SPACING: Missing a blank line after declarations
#755: FILE: samples/nrf_desktop/src/services/hids.c:236:
+	size_t pos = 0;
+	buf[pos] = REPORT_ID_USER_CONFIG;

-:756: ERROR:C99_COMMENTS: do not use C99 // comments
#756: FILE: samples/nrf_desktop/src/services/hids.c:237:
+	pos += 1; // TODO: I dont need this on stack in my struct

-:775: ERROR:C99_COMMENTS: do not use C99 // comments
#775: FILE: samples/nrf_desktop/src/services/hids.c:251:
+	// TODO: have to update data* here

-:794: WARNING:LONG_LINE: line over 80 characters
#794: FILE: samples/nrf_desktop/src/services/hids.c:270:
+			int pos = config_channel_frame_get(rep->data, rep->size, &frame, false);

-:795: WARNING:LINE_SPACING: Missing a blank line after declarations
#795: FILE: samples/nrf_desktop/src/services/hids.c:271:
+			int pos = config_channel_frame_get(rep->data, rep->size, &frame, false);
+			if (pos < 0) {

-:798: WARNING:LONG_LINE_COMMENT: line over 80 characters
#798: FILE: samples/nrf_desktop/src/services/hids.c:274:
+	// TODO: STATUS will be pending from HOST, then SUCCESS from device, possibly (DATA_READY | FORWARD_SUCCESS)

-:798: ERROR:C99_COMMENTS: do not use C99 // comments
#798: FILE: samples/nrf_desktop/src/services/hids.c:274:
+	// TODO: STATUS will be pending from HOST, then SUCCESS from device, possibly (DATA_READY | FORWARD_SUCCESS)

-:801: WARNING:LONG_LINE: line over 80 characters
#801: FILE: samples/nrf_desktop/src/services/hids.c:277:
+				LOG_WRN("Unsupported recipient %" PRIx16, frame.recipient);

-:805: WARNING:LONG_LINE: line over 80 characters
#805: FILE: samples/nrf_desktop/src/services/hids.c:281:
+			struct config_event *event = new_config_event(frame.event_data_len);

-:808: WARNING:LONG_LINE: line over 80 characters
#808: FILE: samples/nrf_desktop/src/services/hids.c:284:
+			memcpy(event->dyndata.data, &(rep->data[pos]), frame.event_data_len);

- total: 5 errors, 11 warnings, 797 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.



```
## Gitlint issues

Commit 3920b2b1ed:
1: UC2 Body does not contain a 'Signed-off-by:' line
1: UC6 Body has no content, should at least have 1 line.
3: B6 Body message is missing

Commit c9e89d8976:
1: UC2 Body does not contain a 'Signed-off-by:' line
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "merge it to branch"
1: UC6 Body has no content, should at least have 1 line.
3: B6 Body message is missing

Commit 8de39e4d3d:
1: UC2 Body does not contain a 'Signed-off-by:' line
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "merge base"
1: UC6 Body has no content, should at least have 1 line.
3: B6 Body message is missing



## Identity/Emails issues

3920b2b1edeca581b892e629d6dd0808ae902e09: author email (Filip Kubicz <filip.kubicz@nordicsemi.no>) needs to match one of the signed-off-by entries.

c9e89d897666e87aec51750d9cde5718cc069af5: author email (Filip Kubicz <filip.kubicz@nordicsemi.no>) needs to match one of the signed-off-by entries.

8de39e4d3d5b24dba27da171520b3dea4ab7e924: author email (Filip Kubicz <filip.kubicz@nordicsemi.no>) needs to match one of the signed-off-by entries.

Qbicz commented:
@pdunaj @MarekPieta, rebased and tested on PCA20041 + PCA10059 (forwarding), PCA20041 BLE to host and USB to host. Can you review?

Note that this does not contain forwarding the data device->dongle->host.

Qbicz commented:
Added fetching of configurable values from the optical sensor - this servers as an example how the bidirectional config channel works.

@pdunaj @MarekPieta can you review?

@kapi-no or @grochu can you review the HIDS service part? This is the first commit here. I modified the feature report handler to indicate to application if this is GATT read or GATT write.

MarekPieta commented on include/bluetooth/services/hids.h:
```diff
@@ -125,11 +125,13 @@ typedef void (*bt_gatt_hids_notif_handler_t) (enum bt_gatt_hids_notif_evt evt);
- * @param rep  Pointer to the report descriptor.
- * @param conn Pointer to Connection Object.
+ * @param rep	Pointer to the report descriptor.
+ * @param conn	Pointer to Connection Object.
```
*connection object?

Qbicz replied:
retained the convention in HIDS docs (Control Point, Connection Object etc)

MarekPieta replied:
ok

MarekPieta commented on include/bluetooth/services/hids.h:
```diff
@@ -125,11 +125,13 @@ typedef void (*bt_gatt_hids_notif_handler_t) (enum bt_gatt_hids_notif_evt evt);
- * @param conn Pointer to Connection Object.
+ * @param rep	Pointer to the report descriptor.
+ * @param conn	Pointer to Connection Object.
+ * @param write	true if handler is called for report write.
```
*True?

rakons replied:
```true``` is the value defined in C whereas ```True``` is not. So it becomes slippy here if you stay grammatically correct.

Qbicz replied:
used `@c true`

MarekPieta commented on samples/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -1010,11 +1055,24 @@ static bool event_handler(const struct event_header *eh)
+			struct config_event *event = cast_config_event(eh);
+
+			if (event->is_fetch) {
+				fetch_config(event->id, event->dyndata.data,
```
Modifying event data here. Should we really do this?
What about other listeners and listeners order?

Qbicz replied:
handled event is now const, I modify only the newly created one

MarekPieta commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -30,6 +30,8 @@ static struct device *usb_dev;
 
 static struct config_channel_frame frame;
 static atomic_t forward_status;
+static u8_t fetched_data[4];
```
name the hardcoded value of 4?

Qbicz replied:
done

MarekPieta commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -40,20 +42,44 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
 					return pos;
 				}
 
+				bool is_fetch = frame.status == FORWARD_STATUS_FETCH;
```
add parenthesis

Qbicz replied:
done

MarekPieta commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -40,20 +42,44 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
 				__ASSERT_NO_MSG(pos <= REPORT_SIZE_USER_CONFIG);
-
-				*len = pos;
+				length = pos;
```
updating local variable just before return 0 ?

Qbicz replied:
This was actually an error. If I had sucessfully modified *len it would introduce too short data length. Removed this non necessary assingment.

MarekPieta commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -91,17 +118,34 @@ static int set_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
 			return -ENOTSUP;
 		}
 
+		bool is_fetch = frame.status == FORWARD_STATUS_FETCH;
```
add parenthesis

Qbicz replied:
done

MarekPieta commented on samples/nrf_desktop/src/services/hids.c:
```diff
@@ -53,6 +53,8 @@ static bool report_enabled[TARGET_REPORT_COUNT][REPORT_MODE_COUNT];
 
 static struct bt_conn *cur_conn;
 
+static u8_t fetched_data[4];
```
Name hardcoded value of 4 ?

Qbicz replied:
done

MarekPieta commented on samples/nrf_desktop/src/services/hids.c:
```diff
@@ -185,38 +187,59 @@ static void feature_report_handler(struct bt_gatt_hids_rep *rep,
+			return;
+		}
+
+		bool is_fetch = frame.status == FORWARD_STATUS_FETCH;
```
add parenthesis ?

Qbicz replied:
done

MarekPieta commented on samples/nrf_desktop/src/services/hids.c:
```diff
@@ -185,38 +187,59 @@ static void feature_report_handler(struct bt_gatt_hids_rep *rep,
+					return;
+				}
+
+				if (fetch_done == true) {
```
`if (fetch_done)`

pdunaj replied:
(also never compare to true)

Qbicz replied:
done

pdunaj commented on samples/nrf_desktop/src/services/hids.c:
```diff
@@ -540,6 +558,19 @@ static bool event_handler(const struct event_header *eh)
 
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_fetch_event(eh)) {
+			struct config_fetch_event *event = cast_config_fetch_event(eh);
```
`const struct config_fetch_event *event`

Qbicz replied:
done

pdunaj commented on scripts/hid_configurator/configurator.py:
```diff
@@ -78,7 +79,12 @@ def create_report(recipient, event_id, event_data):
         event_data_len = 0
 
-    report = struct.pack('<BHBB', REPORT_ID, recipient, event_id, event_data_len)
+    if is_fetch:
```
we don't need data if it is a fetch or do we?

Qbicz replied:
```
>>> struct.pack('')
b''
>>> bool(struct.pack(''))
False
```
But if you prefer to set `event_data = None` I can change it.

pdunaj replied:
but we do not check if data is empty
either assert this is empty or do not process it.

I think it may be better to create two methods: create_set_report and create_fetch_report

Qbicz replied:
done

Qbicz commented on scripts/hid_configurator/configurator.py:
```diff
@@ -106,15 +109,17 @@ def parse_report(report):
 
-def write_feature_with_check(dev, recipient, event_id, event_data):
-    data = create_report(recipient, event_id, event_data)
+def write_feature_with_check(dev, recipient, event_id, event_data, is_fetch):
```
what about `exchange_feature_report()` ?

Qbicz commented on samples/nrf_desktop/src/events/config_event.c:
```diff
@@ -44,3 +49,97 @@ EVENT_TYPE_DEFINE(config_forwarded_event,
+	u8_t buf[REPORT_SIZE_USER_CONFIG] = {0};
+	size_t pos = 0;
+
+	/* BLE HID service requires removal of report ID, according to HOGP_SPEC_V10 */
```
done

Qbicz commented on samples/nrf_desktop/src/events/config_event.c:
```diff
@@ -44,3 +76,96 @@ EVENT_TYPE_DEFINE(config_forwarded_event,
 		  NULL,
 		  NULL);
+
+static int frame_size_check(const struct config_channel_frame *frame,
```
moved to `src/util/config_channel.c`

Qbicz commented on samples/nrf_desktop/src/events/config_event.c:
```diff
@@ -44,3 +76,96 @@ EVENT_TYPE_DEFINE(config_forwarded_event,
+	return 0;
+}
+
+int config_channel_report_parse(const u8_t *buffer, const size_t length,
```
You are right. Removed const from the arguments passed by value

Qbicz commented on samples/nrf_desktop/src/events/config_event.c:
```diff
@@ -44,3 +76,96 @@ EVENT_TYPE_DEFINE(config_forwarded_event,
+
+	__ASSERT_NO_MSG(pos <= length);
+	__ASSERT_NO_MSG(pos <= sizeof(buf));
+	memcpy(buffer, buf, pos);
```
Used `buffer` directly

Qbicz commented on samples/nrf_desktop/src/events/config_event.c:
```diff
@@ -44,3 +76,96 @@ EVENT_TYPE_DEFINE(config_forwarded_event,
+	}
+
+	frame->recipient = sys_get_le16(&buffer[pos]);
+	pos += sizeof(frame->recipient);
```
This is done before, in `frame_size_check()` where sizes of structure members are tested against size of report

Qbicz commented on samples/nrf_desktop/src/events/config_event.c:
```diff
@@ -44,3 +76,96 @@ EVENT_TYPE_DEFINE(config_forwarded_event,
+	buf[pos] = frame->event_data_len;
+	pos += sizeof(frame->event_data_len);
+
+	__ASSERT_NO_MSG(pos <= length);
```
This is done before, in `frame_size_check()` where sizes of structure members are tested against size of report

Qbicz commented on samples/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -752,6 +752,51 @@ static int init(struct config_options *options)
 	return err;
 }
 
+static void fetch_config(const u8_t config_id, u8_t *data, size_t *size)
```
done

Qbicz commented on samples/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -1010,11 +1055,24 @@ static bool event_handler(const struct event_header *eh)
 		if (is_config_event(eh)) {
-			const struct config_event *event =
-				cast_config_event(eh);
+			struct config_event *event = cast_config_event(eh);
```
done

Qbicz commented on samples/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -1010,11 +1055,24 @@ static bool event_handler(const struct event_header *eh)
-				cast_config_event(eh);
+			struct config_event *event = cast_config_event(eh);
+
+			if (event->is_fetch) {
```
Good idea. Fetch request does not need to carry any data, just the ID and the recipient. Also it simplifies checking, as settings subsystem will automatically not receive the fetch request. Done

Qbicz commented on samples/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -1010,11 +1055,24 @@ static bool event_handler(const struct event_header *eh)
+					new_config_fetch_event(event->dyndata.size);
+
+				fetch_event->id = event->id;
+				memcpy(fetch_event->dyndata.data,
```
done

Qbicz commented on samples/nrf_desktop/src/modules/config.c:
```diff
@@ -7,6 +7,7 @@
 #include <stdlib.h>
 
 #include <zephyr.h>
+#include <misc/byteorder.h>
```
relic, removed

Qbicz commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -39,11 +42,44 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
+				}
+
+				bool is_fetch = frame.status == FORWARD_STATUS_FETCH;
+				if (is_fetch) {
```
done

Qbicz commented on samples/nrf_desktop/src/modules/hid_forward.c:
```diff
@@ -282,34 +282,19 @@ static bool event_handler(const struct event_header *eh)
-			pos += sizeof(event_data_len);
-
-			memcpy(&report[pos], event->dyndata.data, event_data_len);
+			memcpy(&report[pos], event->dyndata.data, frame.event_data_len);
```
Now `config_channel_report_fill()` performs the required checks

pdunaj replied:
it seems to be better if this memcpy will be part of the report fill too

Qbicz replied:
done

Qbicz commented:
@pdunaj @MarekPieta I introduced all requested changes. Can you re-review?

joerchan commented:
@Qbicz Can you rebase to solve merge conflict and fix CI build?

kapi-no commented on subsys/bluetooth/services/hids.c:
```diff
@@ -327,7 +327,7 @@ static ssize_t hids_outp_rep_write(struct bt_conn *conn,
 		    .size = rep->size,
 		};
-		rep->handler(&report, conn);
+		rep->handler(&report, conn, true);
```
Please make it consistent and add a callback to the:
```hids_outp_rep_read```

Qbicz replied:
done

kapi-no commented on subsys/bluetooth/services/hids.c:
```diff
@@ -632,7 +640,7 @@ static ssize_t hids_boot_kb_outp_report_write(struct bt_conn *conn,
 		};
 
-		rep->handler(&report, conn);
+		rep->handler(&report, conn, true);
```
Please make it consistent and add a callback to the:
```hids_boot_kb_outp_report_read```

Qbicz replied:
done

kapi-no commented on samples/bluetooth/peripheral_hids_keyboard/src/main.c:
```diff
@@ -234,22 +234,34 @@ static void caps_lock_handler(const struct bt_gatt_hids_rep *rep)
 	char addr[BT_ADDR_LE_STR_LEN];
 
+	if (!write) {
+		printk("Output report read should not happen");
```
Why shouldn't it happen?

I know this is quite unusual, but it is allowed by the spec. You can just state:
"Output report has been read"

kapi-no replied:
Missed "\n" character at the end of string?

Qbicz replied:
done

Qbicz commented:
> @Qbicz Can you rebase to solve merge conflict and fix CI build?

Thanks for a hint. I rebased and fixed conflict.

@kapi-no @grochu I will apply mentioned changes to output reports as well.

MarekPieta commented on samples/nrf_desktop/src/services/hids.c:
```diff
@@ -180,50 +179,64 @@ static void keyboard_leds_handler(struct bt_gatt_hids_rep *rep,
-			return;
-		}
+			event->id = event_id;
+			memcpy(event->dyndata.data, &buffer[pos], event_data_len);
```
Maybe add assert if there is enough data in buffer (at least event_data_len)

MarekPieta replied:
Assert if threre is enough data in buf (based on length)?

MarekPieta replied:
Assert to check if there is enough data in buffer ?

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -39,12 +43,42 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
+						return pos;
+					}
+
+					if (atomic_get(&fetch_done)) {
```
`atomic_get(&fetch_done)` may return a different value then the call in line 58 and lead to inconsistent message.
better to have a local copy of this (i.e. call `atomic_get` only once and operate on the result)

Qbicz replied:
done

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -39,12 +43,42 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
+					return pos;
+				}
+
+				bool is_fetch = (frame.status == FORWARD_STATUS_FETCH);
```
what if we are in the middle of an operation and status from host will change?

Qbicz replied:
Host is required not to do that. It will call set and get report in a sequence, next after completion of the previous. Anyway, here we obtain the status one in "get_report" and then run to completion of this operation.

pdunaj replied:
but if host do it will be react correctly? will the old operation be terminated ? is there any timeout on our side?

pdunaj replied:
?

Qbicz replied:
timeout in firmware added

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -63,73 +97,70 @@ static int set_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
-		event_id = buffer[pos];
-		pos += sizeof(event_id);
+		if (frame.recipient == CONFIG_USB_DEVICE_PID) {
+			atomic_set(&fetch_done, false);
```
maybe a warning if fetch was pending?

Qbicz replied:
added

Qbicz commented:
@pdunaj @MarekPieta I have made the requested changes.

MarekPieta commented on samples/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -1010,11 +1073,31 @@ static bool event_handler(const struct event_header *eh)
+					new_config_fetch_event(data_size);
+
+				fetch_event->id = event->id;
+				fetch_config(fetch_event->id,
```
Initialized to predefined error value in case of fetching unsupported sensor option ?

Qbicz replied:
done

pdunaj replied:
I think what @MarekPieta meant is that there is no way of telling the other board that there was an error when fetching the data.

MarekPieta commented on samples/nrf_desktop/src/util/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
```
2019 ?

Qbicz replied:
done

pdunaj replied:
still different

Qbicz replied:
changed

MarekPieta commented on scripts/hid_configurator/configurator.py:
```diff
@@ -174,29 +222,31 @@ def perform_config(dev, args):
+    if (args.value == 'fetch'):
+        print('Fetch the current value of {} from the firmware'.format(config_name))
+        event_data = b''
+        exchange_feature_report(dev, recipient, event_id, event_data, True)
```
use try here too

Qbicz replied:
done

Qbicz commented on samples/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -752,16 +752,79 @@ static int init(struct config_options *options)
 
+	return;
+error:
+	LOG_WRN("Unsupported sensor option (%" PRIu8 ")", config_id);
```
this log could be removed/changed to assert - we shouldn't enter this function is this is not for us

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -39,12 +43,46 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
+				bool is_fetch = (frame.status == FORWARD_STATUS_FETCH);
+
+				if (is_fetch) {
+					__ASSERT_NO_MSG(frame.event_data_len == 0);
```
Code must never assert on data provided from outside.
Asserts are removed on release build so there will be no check and even if there is a check it's still wrong asmalformed message would be able to crash the device.

Qbicz replied:
done

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -39,12 +43,46 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
+				if (is_fetch) {
+					__ASSERT_NO_MSG(frame.event_data_len == 0);
+
+					if (atomic_get(&fetch_done)) {
```
I think we should check that fetch was for the same item
host can request one thing but subsequent messages can have something else (due to a bug or attempt to crash a device)

Qbicz replied:
done

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -39,12 +43,46 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
+
+					pos = config_channel_report_fill(buffer, length,
+									 &frame, true);
+					if (pos < 0) {
```
this is a common code in both cases

Qbicz replied:
done

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -63,73 +101,76 @@ static int set_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
-		}
+		if (frame.recipient == CONFIG_USB_DEVICE_PID) {
+			if (atomic_get(&fetch_done)) {
+				LOG_WRN("Pending fetched data discarded");
```
This is ok but the other device does not know about the new transaction. It may still work to complete the old request and when it is done the data will be returned as an answer to the new request which is clearly wrong.
A way to fix it would be to add a transaction tag in the request.

Qbicz replied:
only one transaction at a time can be now processed. Also both firmware and host check if the correct event_id and recipient is provided back to host

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -387,6 +428,17 @@ static bool event_handler(const struct event_header *eh)
+		if (is_config_fetch_event(eh)) {
+			struct config_fetch_event *event = cast_config_fetch_event(eh);
+
+			__ASSERT_NO_MSG(event->dyndata.size <= sizeof(fetched_data));
```
if fetched data size comes from outside then assert must not be used

Qbicz replied:
it comes from other part of firmware and should only catch programming mistakes

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -28,7 +29,10 @@ static enum usb_state state;
 static u8_t hid_protocol = HID_PROTOCOL_REPORT;
 static struct device *usb_dev;
 
+static struct config_channel_frame frame;
```
`frame` is both global and local variable which may be confusing and bug prone

Qbicz replied:
done. Actually, it was causing a bug, now fixed

Qbicz commented on samples/nrf_desktop/src/events/config_event.c:
```diff
@@ -44,3 +104,4 @@ EVENT_TYPE_DEFINE(config_forwarded_event,
 		  IS_ENABLED(CONFIG_DESKTOP_INIT_LOG_CONFIG_EVENT),
 		  NULL,
 		  NULL);
+
```
removed

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -28,7 +29,30 @@ static enum usb_state state;
-static atomic_t forward_status;
+static struct config_channel_frame frame;
+static atomic_t status;
+static struct k_delayed_work timeout;
```
I think we should have this in common place for both ble and usb. Reason is not only to avoid duplication.
If we start transactions on both USB and BLE, there seem to be no way of distinguishing which data to direct where and so we may have a problem with formulating a correct reply.

Qbicz replied:
Noted. We need to remember from which transport (and connection) the request came + handle disconnects of BLE/USB

pdunaj commented on samples/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -39,11 +63,57 @@ static int get_report(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
+				u8_t *buffer = *data;
+				int pos;
+
+				// HERE DO NOT PARSE REPORT - it creates a false impression that it is from host.
```
C++ comments are forbidden by CI I think

Qbicz replied:
removed

Qbicz commented:
@pdunaj I have extracted all common logic to config_channel.c and added checks if fetched item is really what was requested.

Can you take a look at the new API, while I add the disconnections handling?

Qbicz commented:
All changes introduced, rebased on master. Can you review @pdunaj @MarekPieta?

joerchan commented on applications/nrf_desktop/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2018 Nordic Semiconductor
+# Copyright (c) 2019 Nordic Semiconductor
```
2018 - 2019

joerchan commented on applications/nrf_desktop/src/modules/Kconfig.config_channel:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2018 Nordic Semiconductor
+# Copyright (c) 2019 Nordic Semiconductor
```
2018 - 2019

joerchan commented on applications/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
2018 - 2019

joerchan commented on applications/nrf_desktop/src/services/hids.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
2018 - 2019

Qbicz replied:
done in multiple places

joerchan commented on applications/nrf_desktop/src/util/config_channel.h:
```diff
@@ -0,0 +1,62 @@
+	struct fetch fetch;
+};
+
+void config_channel_init(struct config_channel_state *cfg_chan);
```
Should we document the functions in this header file?

Qbicz replied:
I will create documentation for this module in subsequent pull request.

Qbicz commented on include/bluetooth/services/hids.h:
```diff
@@ -125,11 +125,13 @@ typedef void (*bt_gatt_hids_notif_handler_t) (enum bt_gatt_hids_notif_evt evt);
-					    struct bt_conn *conn);
+typedef void (*bt_gatt_hids_rep_handler_t) (struct bt_gatt_hids_rep *rep,
+					    struct bt_conn *conn,
+					    const bool write);
```
agree, const not needed when passed by value

Qbicz replied:
done

pdunaj commented on applications/nrf_desktop/src/hw_interface/motion_optical.c:
```diff
@@ -202,14 +207,47 @@ static int init(void)
 }
 
-static void update_config(const u8_t config_id, const u8_t *data, size_t size)
+static inline bool is_my_config_id(u8_t config_id)
```
inline is not needed

Qbicz replied:
done

MarekPieta commented on applications/nrf_desktop/src/modules/usb_state.c:
```diff
@@ -313,6 +254,12 @@ static void device_status(enum usb_dc_status_code cb_status, const u8_t *param)
 		if (new_state == USB_STATE_ACTIVE) {
 			broadcast_subscription_change();
 		}
+
```
Two empty lines

Qbicz replied:
fixed

Qbicz commented:
Hi @wiminordic, do you why this happens during west install in doc build?
`PermissionError: [Errno 13] Permission denied: '/workdir/.local'
`

wiminordic commented:
> Hi @wiminordic, do you why this happens during west install in doc build?
> `PermissionError: [Errno 13] Permission denied: '/workdir/.local' `

Hi, no unfortunately I don't know why it happens. @thst-nordic are you able to help?

Qbicz commented:
@pdunaj @MarekPieta I tested mouse with USB and BLE; and dongle+mouse setup.

@joerchan do you want to re-review?
